### PR TITLE
Wait for guest to settle when obtaining IP address

### PR DESCRIPTION
**Pull request by CheyenneWills**
When obtaining the IP address, try to ping it and retry for a period
of time.  This allows both the guest system and the 'address-source' to
settle in case the 'address-source' has a stale entry.

This problem was discovered with DHCP.  The DHCP service in an ASUS
router holds onto the leases and can have multiple IP addresses leased
to the same MAC address.  When obtaining the IP address, the DNS lookup
would return a stale address before the guest system had actually
obtained an address for the current instance.

### CODE REVIEW

**Comment by meffie** [comment]
Thanks for the patch Cheyenne, but after just a quick look, I'm not sure I want to merge this as is. Minor things:
* this has some trailing whitespace errors
* the lint check fails
* this code normally uses the sh module to run external commands, not os.system

Also, I'm not sure if we want to add an outer retry loop. With this change we nest retry loops, so the max retry time goes from 120 retries to 120 times 120 retries!

File: virt_up/instance.py [hunk 1]
```diff
@@ -590,16 +590,27 @@ def address(self):
...
+                address = self._address_from_ia(source='lease')
+            elif address_source == 'arp':
+                address = self._address_from_arp()
+            elif address_source == 'dns':
+                address = self._address_from_dns()
+            else:
+                raise ValueError(f"Invalid address_source '{address_source}' in instance '{self.name}'.")
+
+            if address is not None:
+                host_up = True if os.system(f"ping -c 2 {address}") is 0 else False
```
Comment (meffie) [inline virt_up/instance.py]: This should be '== 0' not 'is 0'. also we dont need a if:else to set True/False.  e.g.
The rest of this module uses 'sh' to run external commands.

File: virt_up/instance.py [hunk 2]
```diff
@@ -590,16 +590,27 @@ def address(self):
...
+            elif address_source == 'lease':
+                address = self._address_from_ia(source='lease')
+            elif address_source == 'arp':
+                address = self._address_from_arp()
+            elif address_source == 'dns':
+                address = self._address_from_dns()
+            else:
+                raise ValueError(f"Invalid address_source '{address_source}' in instance '{self.name}'.")
+
+            if address is not None:
```
Comment (meffie) [inline virt_up/instance.py]: Should be 'if address:'

File: virt_up/instance.py [hunk 3]
```diff
@@ -590,16 +590,27 @@ def address(self):
...
+                host_up = True if os.system(f"ping -c 2 {address}") is 0 else False
+                if host_up:
+                    break
+
+            if retries > 0:
+                suffix = 'ies' if retries > 1 else 'y'
+                log.debug(f"Waiting for address to settle and ping; {retries} retr{suffix} left.")
+                time.sleep(2)
```
Comment (meffie) [inline virt_up/instance.py]: trailing whitespace

File: virt_up/instance.py [hunk 4]
```diff
@@ -590,16 +590,27 @@ def address(self):
...
+                address = self._address_from_dns()
+            else:
+                raise ValueError(f"Invalid address_source '{address_source}' in instance '{self.name}'.")
+
+            if address is not None:
+                host_up = True if os.system(f"ping -c 2 {address}") is 0 else False
+                if host_up:
+                    break
+
+            if retries > 0:
```
Comment (meffie) [inline virt_up/instance.py:612]: nested retry loops.

**Review by meffie — CHANGES_REQUESTED** [review]

File: virt_up/instance.py (hunk 1 above)
Comment (CheyenneWills) [reply virt_up/instance.py]: ack

File: virt_up/instance.py (hunk 2 above)
Comment (CheyenneWills) [reply virt_up/instance.py]: ack

File: virt_up/instance.py (hunk 3 above)
Comment (CheyenneWills) [reply virt_up/instance.py]: ack

**Comment by meffie** [comment]
superseded by pull request 3

**Comment by meffie** [comment]
superseded by pull request 3

**Pull request closed without merging**